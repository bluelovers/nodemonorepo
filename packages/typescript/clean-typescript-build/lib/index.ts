import {partition} from 'ramda'
import {unlink, lstat, existsSync} from 'fs-extra'
import {traverse} from 'fs-tree-utils'

import {
  Options,
  TargetList,
  Clean
} from './types'

import {
  DEFAULT_DEEP_FUNC,
  DEFAULT_SOURCE_DETECTOR,
  DEFAULT_TARGET_SPECIFIER
} from './contants'

export * from './types'
export * from './contants'

/**
 * Delete all compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns An object
 *   * Property `targets`: An aggregation of `options.listTargets`'s return values
 *   * Property `reports`: An array of reports which tell whether or not deletion succeed
 *   * Property `success`: An array of paths to files which are successfully deleted
 *   * Property `failure`: An array of paths to files which are failed to be deleted
 */
export async function clean (root: string, options?: Options): Promise<Clean.Result> {
  const del = (file: string) => unlink(file).then(onUnlinkSuccess, onUnlinkFailure)
  const r2f = (reports: Clean.Result.ReportList) => reports.map(x => x.file)
  const onUnlinkSuccess = () => ({success: true as true})
  const onUnlinkFailure = (error: any) => ({success: false as false, error})
  const targets = await listAllTargets(root, options)
  const reports = await Promise.all(targets.map(async file => ({file, deletion: await del(file)})))
  const [success, failure] = partition(x => x.deletion.success, reports)
  return {targets, reports, success: r2f(success), failure: r2f(failure)}
}

/**
 * List compiled products corresponding to source files inside `root`
 * @param root Directory that contains all source files
 * @param options Options
 * @param options.deep A function that tells when to go deeper
 * @param options.isSource A function that tells which file is a source file
 * @param options.listTargets A function that lists compiled products corresponding to a source file
 * @returns A promise that resolves a list of target file paths
 */
export async function listAllTargets (
  root: string,
  options: Options = {}
): Promise<TargetList> {
  const {
    deep = DEFAULT_DEEP_FUNC,
    isSource = DEFAULT_SOURCE_DETECTOR,
    listTargets = DEFAULT_TARGET_SPECIFIER
  } = options

  const existed = (await traverse(root, {deep, stat: lstat}))
    .filter(isSource)
    .map(listTargets)
    .reduce((prev, current) => [...prev, ...current], [])
    .filter(item => existsSync(item))

  const files = (await Promise.all(
    existed.map(async item => ({
      item,
      stats: await lstat(item)
    }))
  ))
    .filter(x => x.stats.isFile())
    .map(x => x.item)

  return files
}
