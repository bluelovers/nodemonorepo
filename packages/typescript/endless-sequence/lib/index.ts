import EndlessIterable from './iterable'
import * as applications from './applications'
export {EndlessIterable, applications}
