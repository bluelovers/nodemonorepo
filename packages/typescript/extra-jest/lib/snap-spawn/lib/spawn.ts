import { SpawnSyncOptions, SpawnSyncReturns, spawnSync } from 'child_process'
import ramda from 'ramda'

export type IOData = Buffer | string
export type OptionalIOData = IOData | null | undefined

const fmtNull = <X>(x: X) =>
  x == null ? null : x

const fmtStdIO = (buf: OptionalIOData): string | null => {
  if (buf == null) return null
  const str = String(buf)
  return str.trim() ? `\n${str}\n` : '((EMPTY))'
}

export type Argv = string[]

export interface SpawnReturns {
  readonly status: number | null
  readonly signal: string | null
  readonly error: Error | null
  readonly stdout: string | null
  readonly stderr: string | null
}

export type SpawnFunc = (argv: Argv, options: SpawnSyncOptions) => SpawnSyncReturns<IOData>

export const createSpawnFunc = (command: string): SpawnFunc => ramda.partial(spawnSync, [command])

export function spawn (
  fn: SpawnFunc,
  argv: Argv = [],
  options: SpawnSyncOptions = {}
): SpawnReturns {
  const { status, signal, error, stdout, stderr } = fn(
    argv,
    options
  )

  return {
    status: fmtNull(status),
    signal: fmtNull(signal),
    error: fmtNull(error),
    stdout: fmtStdIO(stdout),
    stderr: fmtStdIO(stderr)
  }
}

export namespace spawn {
  export const withCommand = (
    command: string,
    argv?: Argv,
    options?: SpawnSyncOptions
  ) => spawn(createSpawnFunc(command), argv, options)
}

export default spawn
