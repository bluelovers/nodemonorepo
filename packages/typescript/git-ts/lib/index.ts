import Repository from './repo'
export {Repository}
