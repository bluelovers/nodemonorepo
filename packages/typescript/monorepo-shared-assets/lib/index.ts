import * as group from './group'
import * as iter from './iter'
import * as asyncIter from './async-iter'

export * from './.types'

export {
  group,
  iter,
  asyncIter
}
