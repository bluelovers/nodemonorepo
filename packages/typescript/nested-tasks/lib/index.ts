import listManifestFiles from './list-manifest-files'
export {listManifestFiles}
