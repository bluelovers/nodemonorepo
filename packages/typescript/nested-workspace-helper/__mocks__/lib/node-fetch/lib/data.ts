import {
  PackageRegistry,
  PackageVersionRegistry,
  RegistryPackageVersionSet
} from '../../../../lib/types'

const pkgVers = require('../data/pkg-vers.yaml') as Readonly<{
  [name: string]: ReadonlyArray<string>
}>

export {
  PackageRegistry,
  PackageVersionRegistry,
  RegistryPackageVersionSet
}

export interface PackageSet {
  readonly [name: string]: PackageRegistry
}

export const packages = Object
  .entries(pkgVers)
  .map(([name, versions]) => ({
    name,
    versions: versions
      .map(
        version =>
          ({name, version} as PackageVersionRegistry)
      )
      .reduce(
        (prev, current) => Object.assign(prev, {[current.version]: current}),
        {} as RegistryPackageVersionSet
      )
  } as PackageRegistry))
  .reduce(
    (prev, current) => Object.assign(prev, {[current.name]: current}),
    {} as PackageSet
  )
