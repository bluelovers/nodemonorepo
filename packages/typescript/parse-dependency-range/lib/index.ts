export * from './types'
export * from './constants'
export * from './functions'

// workaround a stupid bug during test:
//   TypeError: Cannot read property 'Semver' of undefined
//
//     4 |
//     5 | const samples: ReadonlyArray<Sample> = [
//   > 6 |   [Type.Semver, [
//       |         ^
//     7 |     '0.1.2',
//     8 |     '=0.1.2',
//     9 |     '~0.1.2',
//
//   at Object.<anonymous> (packages/typescript/parse-dependency-range/test/lib.test/functions.test/parse.test/index.test.ts:6:9)
export { Type } from './types'
