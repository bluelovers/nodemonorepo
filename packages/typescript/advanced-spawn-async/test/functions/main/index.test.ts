import {ChildProcess} from 'child_process'
import spawn, {main, SpawnFactory, SpawnError} from '../../../index'
import * as data from '../../.lib/data'

type Info = SpawnFactory.TerminationInformation<ChildProcess>
type SError = SpawnError<ChildProcess, Info>

const redacted = '[REDACTED]'

const sanitize = (
  info: Info,
  {
    args = false
  }: {
    args?: boolean
  } = {}
) => ({
  ...info,
  process: {
    constructor: {
      name: info.process.constructor.name
    }
  },
  args: args ? info.args : redacted as typeof redacted
})

const sanitizeError = (error: SError) => ({
  ...error,
  message: redacted as typeof redacted,
  stack: redacted as typeof redacted,
  info: sanitize(error.info)
})

// beforeEach(() => {
//   jest.setTimeout(131072)
// })

// afterEach(() => {
//   jest.setTimeout(5000)
// })

it('export main as default', () => {
  expect(spawn).toBe(main)
})

describe('when process successfully terminated', () => {
  describe('with minimal arguments', () => {
    const factory = spawn('echo')

    it('process', () => {
      expect({
        ...factory,
        process: {
          constructor: {
            name: factory.process.constructor.name
          }
        }
      }).toMatchSnapshot()
    })

    describe('close promise', () => {
      const {close, onclose} = factory

      it('close is onclose()', () => {
        expect(close).toBe(onclose())
      })

      it('close->process is factory.process', async () => {
        expect((await close).process).toBe(factory.process)
      })

      it('matches snapshot', async () => {
        expect(sanitize(await close, {args: true})).toMatchSnapshot()
      })
    })

    describe('exit promise', () => {
      const {exit, onexit} = factory

      it('exit is onexit()', () => {
        expect(exit).toBe(onexit())
      })

      it('exit->process is factory.process', async () => {
        expect((await exit).process).toBe(factory.process)
      })

      it('matches snapshot', async () => {
        expect(sanitize(await exit, {args: true})).toMatchSnapshot()
      })
    })
  })

  describe('with full arguments', () => {
    const process = spawn('echo', ['hello', 'world'], {env: {HELLO: 'WORLD'}})

    it('close promise', async () => {
      expect(sanitize(await process.close, {args: true})).toMatchSnapshot()
    })

    it('exit promise', async () => {
      expect(sanitize(await process.exit, {args: true})).toMatchSnapshot()
    })
  })

  describe('with stdout and stderr', () => {
    const factory = spawn('node', [data.bothStdoutStderr], {env: {HELLO: 'WORLD'}})

    it('close promise', async () => {
      expect(sanitize(await factory.close)).toMatchSnapshot
    })


    it('exit promise', async () => {
      expect(sanitize(await factory.exit)).toMatchSnapshot
    })
  })

  describe('with specified stdin', () => {
    const {process, close, exit} = spawn('bash')

    it('close promise', async () => {
      expect(sanitize(await close, {args: true})).toMatchSnapshot
    })

    it('exit promise', async () => {
      expect(sanitize(await exit, {args: true})).toMatchSnapshot
    })

    process.stdin.write('echo stdin foo\n')
    process.stdin.write('echo stderr foo 1>&2\n')
    process.stdin.write('echo stdin bar\n')
    process.stdin.write('echo stderr bar 1>&2\n')
    process.stdin.end('exit 0\n')
  })
})

describe('when process terminated with non-zero status code', () => {
  const factory = spawn('node', [data.withNonZeroStatus], {env: {HELLO: 'WORLD'}})

  it('close promise', async () => {
    const result = await factory.close.then(
      () => Promise.reject(new Error('factory.close should not resolve')),
      error => Object.assign(error, {args: '[REDACTED]'})
    )

    expect(sanitizeError(result)).toMatchSnapshot()
  })

  it('exit promise', async () => {
    const result = await factory.exit.then(
      () => Promise.reject(new Error('factory.close should not resolve')),
      error => Object.assign(error, {args: '[REDACTED]'})
    )

    expect(sanitizeError(result)).toMatchSnapshot()
  })
})
