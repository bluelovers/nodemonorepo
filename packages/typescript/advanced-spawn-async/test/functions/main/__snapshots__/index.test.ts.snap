// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when process successfully terminated with minimal arguments close promise matches snapshot 1`] = `
Object {
  "args": Array [],
  "command": "echo",
  "options": Object {},
  "output": "
",
  "process": Object {
    "constructor": Object {
      "name": "ChildProcess",
    },
  },
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "
",
}
`;

exports[`when process successfully terminated with minimal arguments exit promise matches snapshot 1`] = `
Object {
  "args": Array [],
  "command": "echo",
  "options": Object {},
  "output": "
",
  "process": Object {
    "constructor": Object {
      "name": "ChildProcess",
    },
  },
  "signal": null,
  "status": 0,
  "stderr": "",
  "stdout": "
",
}
`;

exports[`when process successfully terminated with minimal arguments process 1`] = `
Object {
  "close": Promise {},
  "exit": Promise {},
  "onclose": [Function],
  "onexit": [Function],
  "process": Object {
    "constructor": Object {
      "name": "ChildProcess",
    },
  },
}
`;
